// @aws-amplify/storage/lib/common/StorageUtils.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getStorageUserAgentValue = exports.isBlob = exports.isFile = exports.dispatchStorageEvent = exports.byteLength = void 0;
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
var core_1 = require("@aws-amplify/core");
var StorageConstants_1 = require("./StorageConstants");
var byteLength = function (x) {
    if (typeof x === 'string') {
        return x.length;
    }
    else if (isArrayBuffer(x)) {
        return x.byteLength;
    }
    else if ((0, exports.isBlob)(x)) {
        return x.size;
    }
    else {
        throw new Error('Cannot determine byte length of ' + x);
    }
};
exports.byteLength = byteLength;
var dispatchStorageEvent = function (track, event, attrs, metrics, message) {
    if (track) {
        var data = { attrs: attrs };
        if (metrics) {
            data['metrics'] = metrics;
        }
        core_1.Hub.dispatch('storage', {
            event: event,
            data: data,
            message: message,
        }, 'Storage', StorageConstants_1.AMPLIFY_SYMBOL);
    }
};
exports.dispatchStorageEvent = dispatchStorageEvent;
var isFile = function (x) {
    return typeof x !== 'undefined' && x instanceof File;
};
exports.isFile = isFile;
var isBlob = function (x) {
    return typeof x !== 'undefined' && x instanceof Blob;
};
exports.isBlob = isBlob;
var getStorageUserAgentValue = function (action, customUserAgentDetails) {
    return (0, core_1.getAmplifyUserAgent)(__assign({ category: core_1.Category.Storage, action: action }, customUserAgentDetails));
};
exports.getStorageUserAgentValue = getStorageUserAgentValue;
var isArrayBuffer = function (x) {
    return typeof x !== 'undefined' && x instanceof ArrayBuffer;
};
